File: golang/agents/loop.go
package agents

// Loop — a live execution of an Agent, and the completion gate that stops it
// claiming `done` too early. A PROTOTYPE layer over the shipped §8 client:
// nothing here changes existing behaviour.
//
// The placement law this encodes:
//
//	Spec (the harness) answers "MAY it?"      — capability, ceilings.  Per problem.
//	Run                answers "with WHAT?"   — model for this call.   Per call.
//	Loop               answers "DID it?"      — status, turns.         Observed.
//	none of them       answers "is it RIGHT?" — a tool, skill or agent.
//
// So Loop takes no options: it is read, not configured.

import (
	"context"
	"fmt"
	"strings"

	tn "github.com/muthuishere/toolnexus/golang"
)

// guardedHooks compiles a Spec's guardrails into one BeforeTool with
// FIRST-DENY-WINS, composed ahead of any hook the Spec already set.
func guardedHooks(sp Spec) *tn.Hooks {
	if len(sp.Guardrails) == 0 {
		return sp.Hooks
	}
	prior := func(context.Context, tn.BeforeToolEvent) (*tn.ToolOverride, error) { return nil, nil }
	merged := tn.Hooks{}
	if sp.Hooks != nil {
		merged = *sp.Hooks
		if sp.Hooks.BeforeTool != nil {
			prior = sp.Hooks.BeforeTool
		}
	}
	rails := sp.Guardrails
	merged.BeforeTool = func(ctx context.Context, ev tn.BeforeToolEvent) (*tn.ToolOverride, error) {
		for _, g := range rails {
			if v := g(ev); v != "" && v != "allow" {
				return &tn.ToolOverride{Result: &tn.ToolResult{
					Output: "denied: " + v, IsError: true}}, nil
			}
		}
		return prior(ctx, ev)
	}
	return &merged
}

// Loop is a live execution of an Agent. Its only verbs are Run and Resume.
type Loop struct {
	agent   *Agent
	opts    tn.ClientOptions // kept so a per-run Model override can rebuild the client
	toolkit *tn.Toolkit
	history []any
	calls   []tn.ToolCall // accumulated across attempts — see Run
	turns   int
	status  string
}

// RunOpts is what varies PER CALL. Model is here, not on the Spec's default and
// not on the Loop, so the same conversation may change model between turns.
type RunOpts struct {
	// Model overrides the agent's model for this call only. "" ⇒ the agent's.
	Model string
}

// Outcome is what a Run reports. Status reuses the SHIPPED vocabulary — no new
// status strings are minted (SPEC.md pins TaskStatus identical across ports).
type Outcome struct {
	Text      string
	Status    string // done | incomplete | pending | error
	StoppedBy string // named whenever Status != "done" — never a silent stop
	Attempts  int
	Turns     int
	Result    tn.RunResult
}

// Loop opens a live execution for an agent. It takes the CLIENT OPTIONS rather
// than a built client, because RunOpts.Model must be able to override the model
// for a single call — and the model is fixed when a client is constructed.
func (a *Agent) Loop(opts tn.ClientOptions, tk *tn.Toolkit) *Loop {
	return &Loop{agent: a, opts: opts, toolkit: tk, status: "idle"}
}

// clientFor returns the client for this call, applying a per-run Model override
// via RequestParams (`model` is NOT in the forbidden set — client.go forbids only
// messages/tools/stream — and the merge reaches the wire; spiked).
func (l *Loop) clientFor(o RunOpts) *tn.Client {
	if o.Model == "" {
		return tn.CreateClient(l.opts)
	}
	opts := l.opts
	rp := map[string]any{}
	for k, v := range opts.RequestParams {
		rp[k] = v
	}
	rp["model"] = o.Model
	opts.RequestParams = rp
	return tn.CreateClient(opts)
}

// Status is observed, never set by the caller.
func (l *Loop) Status() string { return l.status }

// Turns is the model round trips this loop has spent.
func (l *Loop) Turns() int { return l.turns }

// Run executes one request. When the agent's Spec declares a Completion, the
// gate runs exactly where the loop would otherwise have returned `done`.
func (l *Loop) Run(ctx context.Context, prompt string, opts RunOpts) (Outcome, error) {
	client := l.clientFor(opts)
	l.status = "running"

	attempts := 0
	// The gate lives in runGated, NOT here. An inline second copy is how this port
	// came to report a completion stop without setting Result.Limit, while the
	// runtime path set it — one behaviour, two implementations, one of them wrong.
	r, err := runGated(func(p string) (tn.RunResult, error) {
		attempts++
		out, err := client.RunWithHistory(ctx, p, l.toolkit, l.history)
		if err != nil {
			return out, err
		}
		l.turns += out.Turns
		l.history = out.Messages
		return out, nil
	}, prompt, l.agent.Spec.Completion)

	if err != nil {
		l.status = "error"
		return Outcome{Status: "error", StoppedBy: err.Error(), Attempts: attempts, Turns: l.turns}, err
	}

	if r.Status != "" && r.Status != "done" {
		l.status = r.Status
		stoppedBy := "run reported " + r.Status
		if r.Limit == "completion" {
			stoppedBy = r.Text
		}
		return Outcome{Text: r.Text, Status: r.Status, StoppedBy: stoppedBy,
			Attempts: attempts, Turns: l.turns, Result: r}, nil
	}

	l.status = "idle"
	return Outcome{Text: r.Text, Status: "done", Attempts: attempts, Turns: l.turns, Result: r}, nil
}

// AllTodosDone is the built-in completion verifier. It reads the SHIPPED
// `todowrite` builtin's result metadata and requires every item to be checked.
//
// Structural, not domain: it counts unchecked boxes and never learns what a todo
// means, so the loop stays domain-blind. No plan declared ⇒ nothing to verify ⇒
// pass, so the gate never punishes an agent that does not use the builtin.
func AllTodosDone(r tn.RunResult) (bool, string) {
	for i := len(r.ToolCalls) - 1; i >= 0; i-- {
		if r.ToolCalls[i].Name != "todowrite" {
			continue
		}
		raw, ok := r.ToolCalls[i].Metadata["todos"].([]any)
		if !ok {
			return true, ""
		}
		var open []string
		for _, it := range raw {
			m, _ := it.(map[string]any)
			if done, _ := m["completed"].(bool); !done {
				txt, _ := m["text"].(string)
				open = append(open, txt)
			}
		}
		if len(open) > 0 {
			return false, fmt.Sprintf("%d item(s) still open: %s", len(open), strings.Join(open, "; "))
		}
		return true, ""
	}
	return true, ""
}

// runGated wraps a client run with the completion gate. It is the SHARED
// implementation used by both the standalone Loop and the §7D runtime turn, so
// a delegated child gets exactly the same guarantee as a directly-driven one.
//
// Rule 2 in force: a run that is `pending` (suspended on a human) or otherwise
// non-done already carries its own reason, so the gate never re-judges it. That
// keeps `pending` and `incomplete` distinct — the caller can always tell whether
// it owes an Answer or a fix.
// ask runs ONE attempt. Taking a closure rather than a client is what lets the
// standalone Loop and the runtime turn share this function instead of keeping two
// copies of the six rules — which is how the two drifted in the first place.
type ask func(prompt string) (tn.RunResult, error)

func runGated(a ask, prompt string, comp *Completion) (tn.RunResult, error) {
	if comp == nil {
		return a(prompt)
	}
	maxAttempts := comp.MaxAttempts
	if maxAttempts < 1 {
		return tn.RunResult{}, fmt.Errorf("toolnexus: Completion.MaxAttempts must be >= 1")
	}
	if comp.Verify == nil {
		return tn.RunResult{}, fmt.Errorf("toolnexus: Completion.Verify is required")
	}
	var acc []tn.ToolCall
	var last tn.RunResult
	var reason string
	for n := 1; n <= maxAttempts; n++ {
		p := prompt
		if n > 1 {
			p = "Your work did not verify: " + reason + ". Fix it and finish."
		}
		r, err := a(p)
		if err != nil {
			return r, err
		}
		// The gate judges the accumulated work, so an agent cannot escape it by
		// declining to re-declare its plan on a retry.
		acc = append(acc, r.ToolCalls...)
		r.ToolCalls = acc
		last = r
		if r.Status != "" && r.Status != "done" {
			// The run stopped for its own reason (suspension, budget). If the gate
			// was mid-retry, the caller must learn BOTH — otherwise a budget stop
			// masks the verification failure and they never see why it was looping.
			// Spiked: without this the caller reads only "hit maxTurns".
			if reason != "" && r.Status != "pending" {
				r.Text = fmt.Sprintf("%s [while verifying: attempt %d last failed: %s]",
					r.Text, n, reason)
			}
			return r, nil
		}
		ok, why := comp.Verify(r)
		if ok {
			return r, nil
		}
		reason = why
	}
	// Structured, not prose: `Limit` is how a caller (and the §7D runtime) tells
	// WHICH limit stopped the run. Text carries the human reason.
	last.Status = "incomplete"
	last.Limit = "completion"
	last.Text = fmt.Sprintf("completion.verify failed %d×: %s", maxAttempts, reason)
	return last, nil
}

File: js/src/agents/loop.ts
/**
 * Loop — a live execution of an Agent, and the completion gate that stops it
 * claiming `done` too early. A layer over the shipped §8 client: nothing here
 * changes existing behaviour.
 *
 * The placement law this encodes:
 *
 *   AgentSpec (the harness) answers "MAY it?"    — capability, ceilings. Per problem.
 *   RunOptions             answers "with WHAT?"  — model for this call.  Per call.
 *   Loop                   answers "DID it?"     — status, turns.        Observed.
 *   none of them           answers "is it RIGHT?"— a tool, skill or agent.
 *
 * So Loop takes no options: it is read, not configured.
 */
import type { PromptInput } from "../content.js"
import { createClient, type ClientOptions, type Hooks, type RunResult } from "../client.js"
import type { Toolkit } from "../toolkit.js"
import type { Agent } from "./agent.js"

/** A POLICY check on a tool call — "may it?", never "is it right?".
 *  Return `"allow"` (or nothing) to permit; any other string DENIES with that reason. */
export type Guardrail = (ev: {
  name: string
  args: Record<string, unknown>
  id?: string
  turn: number
}) => string | undefined | void

/** The verdict a completion verifier returns. */
export interface Verdict {
  ok: boolean
  reason?: string
}

/** The gate that stops an agent claiming `done` before its work verifies. */
export interface Completion {
  /** Judges the run. Receives the tool calls ACCUMULATED across attempts. */
  verify: (result: RunResult) => Verdict | Promise<Verdict>
  /** REQUIRED. An unbounded verify loop is a denial-of-service on the caller's bill. */
  maxAttempts: number
}

/** What varies PER CALL. `model` is here — not on the spec's default and not on the
 *  Loop — so the same conversation may change model between turns. */
export interface LoopRunOptions {
  /** Overrides the agent's model for this call only. Omitted ⇒ the agent's. */
  model?: string
}

/** What a run reports. `status` reuses the SHIPPED vocabulary — no new status
 *  strings are minted (SPEC.md pins TaskStatus identical across ports). */
export interface Outcome {
  text: string
  status: "done" | "incomplete" | "pending" | "error"
  /** Named whenever `status` is not `done` — never a silent stop. */
  stoppedBy?: string
  attempts: number
  turns: number
  result?: RunResult
}

/**
 * Compiles a spec's guardrails into one `beforeTool` with FIRST-DENY-WINS,
 * composed ahead of any hook the spec already set. No guardrails ⇒ the spec's
 * hooks are returned untouched, so absent is byte-identical.
 */
export function guardedHooks(guardrails: Guardrail[] | undefined, hooks: Hooks | undefined): Hooks | undefined {
  if (!guardrails || guardrails.length === 0) return hooks
  const prior = hooks?.beforeTool
  return {
    ...(hooks ?? {}),
    beforeTool: async (ev) => {
      for (const g of guardrails) {
        const verdict = g(ev)
        // A guardrail is SYNCHRONOUS by contract. An async one returns a Promise,
        // which is truthy and !== "allow", so it would silently DENY every call
        // with `denied: [object Promise]`. Fail loudly instead, and say where
        // asynchronous work belongs.
        if (verdict != null && typeof verdict !== "string") {
          const what = typeof (verdict as { then?: unknown }).then === "function" ? "a Promise" : typeof verdict
          throw new TypeError(
            `guardrail returned ${what}: a guardrail must return a string synchronously ` +
              `("" or "allow" to permit, any other string to deny). ` +
              `Do asynchronous work in hooks.beforeTool, which is awaited.`,
          )
        }
        if (verdict && verdict !== "allow") {
          return { result: { output: `denied: ${verdict}`, isError: true } }
        }
      }
      return prior ? prior(ev) : undefined
    },
  }
}

/**
 * The built-in completion verifier. Reads the SHIPPED `todowrite` builtin's
 * result metadata and requires every item to be checked.
 *
 * Structural, not domain: it counts unchecked boxes and never learns what a todo
 * means, so the loop stays domain-blind. No plan declared ⇒ nothing to verify ⇒
 * pass, so the gate never punishes an agent that does not use the builtin.
 */
export function allTodosDone(result: RunResult): Verdict {
  for (let i = result.toolCalls.length - 1; i >= 0; i--) {
    if (result.toolCalls[i].name !== "todowrite") continue
    const todos = result.toolCalls[i].metadata?.todos
    if (!Array.isArray(todos)) return { ok: true }
    const open = todos
      .filter((t: any) => !t?.completed)
      .map((t: any) => String(t?.text ?? ""))
    return open.length > 0
      ? { ok: false, reason: `${open.length} item(s) still open: ${open.join("; ")}` }
      : { ok: true }
  }
  return { ok: true }
}

/**
 * Wraps a client run with the completion gate. SHARED by the standalone Loop and
 * the §7D runtime turn, so a delegated child gets exactly the same guarantee as a
 * directly-driven one.
 *
 * Rule 2 in force: a run that is `pending` (suspended on a human) or otherwise
 * non-done already carries its own reason, so the gate never re-judges it. That
 * keeps `pending` and `incomplete` distinct — the caller can always tell whether
 * it owes an Answer or a fix.
 */
export async function runGated(
  ask: (prompt: PromptInput) => Promise<RunResult>,
  prompt: PromptInput,
  completion: Completion | undefined,
): Promise<RunResult> {
  if (!completion) return ask(prompt)
  const { maxAttempts } = completion
  if (!Number.isInteger(maxAttempts) || maxAttempts < 1) {
    throw new Error("toolnexus: completion.maxAttempts must be an integer >= 1")
  }
  if (typeof completion.verify !== "function") {
    throw new Error("toolnexus: completion.verify is required")
  }

  let accumulated: RunResult["toolCalls"] = []
  let last: RunResult | undefined
  let reason = ""
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    const p = attempt === 1 ? prompt : `Your work did not verify: ${reason}. Fix it and finish.`
    const r = await ask(p)
    // The gate judges the ACCUMULATED work, so an agent cannot escape it by
    // declining to re-declare its plan on a retry.
    accumulated = [...accumulated, ...r.toolCalls]
    const judged: RunResult = { ...r, toolCalls: accumulated }
    last = judged

    if (r.status && r.status !== "done") {
      // The run stopped for its own reason (suspension, budget). If the gate was
      // mid-retry the caller must learn BOTH — otherwise a budget stop masks the
      // verification failure and they never see why it was looping.
      if (reason && r.status !== "pending") {
        judged.text = `${r.text} [while verifying: attempt ${attempt} last failed: ${reason}]`
      }
      return judged
    }

    const verdict = await completion.verify(judged)
    if (verdict.ok) return judged
    reason = verdict.reason ?? "unspecified"
  }

  // Structured, not prose: `limit` is how a caller (and the §7D runtime) tells
  // WHICH limit stopped the run. `text` carries the human reason.
  return {
    ...(last as RunResult),
    status: "incomplete",
    limit: "completion",
    text: `completion.verify failed ${maxAttempts}×: ${reason}`,
  }
}

/**
 * The spec fields a Loop drive CANNOT honour, named rather than dropped silently.
 *
 * A `Loop` is one client run: it has no runtime, so it has no handle tree to spawn a `team`
 * into, no §10 suspension ledger for `waitFor`, and no per-agent metric sink routing. Those
 * fields are honoured by `Agent.run` / `AgentRuntime` only. `uses.tools` is not honoured either
 * — a Loop is handed its toolkit by the caller.
 *
 * Returns a FIXED CANONICAL vocabulary — `"tools"`, `"team"`, `"waitFor"`, `"onMetric"` —
 * identical in all seven ports, exactly as the `limit` strings are, and NOT this language's own
 * spelling of the field. Empty ⇒ nothing is lost.
 * Additive and advisory: a Loop never refuses to run because of it (that would be breaking).
 */
export function loopUnsupported(spec: {
  uses?: { tools?: unknown[] }
  team?: unknown[]
  waitFor?: unknown
  onMetric?: unknown
}): string[] {
  const out: string[] = []
  if (spec.uses?.tools && spec.uses.tools.length > 0) out.push("tools")
  if (spec.team && spec.team.length > 0) out.push("team")
  if (spec.waitFor) out.push("waitFor")
  if (spec.onMetric) out.push("onMetric")
  return out
}

/** A live execution of an Agent. Its only verbs are `run` and reading state. */
export class Loop {
  #status: Outcome["status"] | "idle" | "running" = "idle"
  #turns = 0

  constructor(
    private readonly agent: Agent,
    /** CLIENT OPTIONS rather than a built client, because a per-call `model`
     *  override must be able to change the model — which is fixed at construction. */
    private readonly options: ClientOptions,
    private readonly toolkit: Toolkit,
  ) {}

  /** Observed, never set by the caller. */
  get status(): string {
    return this.#status
  }

  /** Model round trips this loop has spent. */
  get turns(): number {
    return this.#turns
  }

  async run(prompt: PromptInput, opts: LoopRunOptions = {}): Promise<Outcome> {
    const completion = this.agent.spec.completion
    const client = createClient(this.#clientOptions(opts))
    this.#status = "running"

    let attempts = 0
    let history: any[] = []
    const ask = async (p: PromptInput): Promise<RunResult> => {
      attempts++
      const r = await client.run(p, { toolkit: this.toolkit, history })
      this.#turns += r.turns
      history = r.messages
      return r
    }

    try {
      const r = await runGated(ask, prompt, completion)
      if (r.status && r.status !== "done") {
        this.#status = r.status as Outcome["status"]
        const stoppedBy =
          r.limit === "completion" ? r.text : `run reported ${r.status}`
        return { text: r.text, status: r.status as Outcome["status"], stoppedBy, attempts, turns: this.#turns, result: r }
      }
      this.#status = "idle"
      return { text: r.text, status: "done", attempts, turns: this.#turns, result: r }
    } catch (e) {
      this.#status = "error"
      throw e
    }
  }

  /** Applies a per-call model override via `requestParams` (`model` is not in the
   *  forbidden set — the client forbids only messages/tools/stream). */
  #clientOptions(opts: LoopRunOptions): ClientOptions {
    const spec = this.agent.spec
    const base: ClientOptions = {
      ...this.options,
      // CALLER-WINS (ADR 0024 / issue #87). A `systemPrompt` the caller put on the client
      // options is an explicit instruction for THIS drive; the soul is the agent's default.
      // js used to let the soul override it, which python and csharp never did.
      systemPrompt: this.options.systemPrompt ?? spec.soul,
      // The spec's model and turn ceiling become Loop DEFAULTS — the same spec drives both
      // doors. `ClientOptions.model` is REQUIRED in js, so "the caller did not choose" is
      // spelled the way the runtime already spells it: absent, or the `"inherit"` sentinel.
      model: !this.options.model || this.options.model === "inherit"
        ? (spec.model && spec.model !== "inherit" ? spec.model : this.options.model)
        : this.options.model,
      maxTurns: this.options.maxTurns ?? spec.budget?.maxTurns,
      hooks: guardedHooks(spec.guardrails, spec.hooks ?? this.options.hooks),
    }
    if (!opts.model) return base
    return { ...base, requestParams: { ...(base.requestParams ?? {}), model: opts.model } }
  }
}
